app/models/compute_resource.rb

```diff
@@ -52,18 +41,31 @@ class ComputeResource < ActiveRecord::Base
     end
   }
 
-  def self.register_provider(provider)
-    name = provider.name.split('::').last
-    return if supported_providers.values.include?(provider) || supported_providers.keys.include?(name)
-    supported_providers[name] = provider.name
+  def self.supported_providers
+    {
+      'Libvirt'   => 'Foreman::Model::Libvirt',
+      'Ovirt'     => 'Foreman::Model::Ovirt',
+      'EC2'       => 'Foreman::Model::EC2',
+      'Vmware'    => 'Foreman::Model::Vmware',
+      'Openstack' => 'Foreman::Model::Openstack',
+      'Rackspace' => 'Foreman::Model::Rackspace',
+      'GCE'       => 'Foreman::Model::GCE',
+    }
+  end
+
+  def self.registered_providers
+    Foreman::Plugin.all.map(&:compute_resources).inject({}) do |prov_hash,providers|
+      providers.each { |provider| prov_hash.update(provider.name.split('::').last => provider.name) }
+      prov_hash
+    end
   end
 
   def self.providers
-    supported_providers.reject { |p,c| !SETTINGS[p.downcase.to_sym] }.keys
+    (supported_providers.reject { |p,c| !SETTINGS[p.downcase.to_sym] }.keys + registered_providers.keys).uniq
```

Comment:
The settings here aren't real settings, it's used for convenient global storage.  It's been used forever in Foreman in various forms - config/application.rb sets these values depending on whether it can load the relevant Bundler groups and Fog providers.  For plugins (prior to this PR) the value of SETTINGS[:docker] etc is forced to true in the plugin registration.

I don't think it's possible to disable a compute resource using SETTINGS today.  I agree that being able to disable plugins would be a useful feature though, but this isn't it.

As an aside, with Fog nearly entirely modularised, we can soon simplify this to checking whether the provider is registered in Fog - see #3370 for a small bit of progress in refactoring this.